modules/renter/download.go

```diff
@@ -256,22 +207,91 @@ func (r *Renter) Download(path, destination string) error {
 	}
 	defer f.Close()
 
-	// Create the download object.
-	d := file.newDownload(hosts, destination)
+	// A loop that will iterate until the download is complete.
+	var chunkIndex, received uint64
+	for {
+		// copy file contracts
+		file.mu.RLock()
+		contracts := make([]fileContract, 0, len(file.contracts))
+		for _, c := range file.contracts {
+			contracts = append(contracts, c)
+		}
+		file.mu.RUnlock()
+		if len(contracts) < file.erasureCode.MinPieces() {
+			return fmt.Errorf("contracts could not be located for this file - file may not be recoverable - needed %v, got %v", file.erasureCode.MinPieces(), len(contracts))
+		}
 
-	// Add the download to the download queue.
-	lockID = r.mu.Lock()
-	r.downloadQueue = append(r.downloadQueue, d)
-	r.mu.Unlock(lockID)
+		// interrupt upload loop
+		lockID = r.mu.Lock()
+		r.downloading = true
+		uploading := r.uploading
+		r.mu.Unlock(lockID)
+		defer func() {
+			lockID = r.mu.Lock()
+			r.downloading = false
+			r.mu.Unlock(lockID)
+		}()
+		// wait up to 60 minutes for upload loop to exit
+		timeout := time.Now().Add(60 * time.Minute)
+		for uploading && time.Now().Before(timeout) {
+			time.Sleep(time.Second)
+			lockID = r.mu.RLock()
+			uploading = r.uploading
+			r.mu.RUnlock(lockID)
+		}
+		if uploading {
+			return errors.New("timed out waiting for uploads to finish")
+		}
 
-	// Perform download.
-	err = d.run(f)
-	if err != nil {
-		// File could not be downloaded; delete the copy on disk.
-		os.Remove(destination)
-		return err
-	}
+		// Grab a set of hosts and attempt a download.
+		err := func() error {
+			// Initiate connections to each host.
+			var hosts []fetcher
+			var errs []string
+			for _, c := range file.contracts {
+				d, err := r.hostContractor.Downloader(c.ID)
+				if err != nil {
+					errs = append(errs, fmt.Sprintf("\t%v: %v", c.IP, err))
+					continue
+				}
+				defer d.Close()
+				hosts = append(hosts, newHostFetcher(d, c.Pieces, file.masterKey))
+			}
+			if len(hosts) < file.erasureCode.MinPieces() {
+				return errors.New("could not connect to enough hosts:\n" + strings.Join(errs, "\n"))
+			}
+			// Check that this host set is sufficient to download the file.
+			err := checkHosts(hosts, file.erasureCode.MinPieces(), file.numChunks())
+			if err != nil {
+				return err
+			}
 
+			// We have a new set of hosts, update the downloader to reflect that.
+			lockID := r.mu.Lock()
+			// TODO: Not quite sure how thread safe this is.
+			d.hosts = hosts
+			r.mu.Unlock(lockID)
+			defer func() {
+				// Clear the set of hosts from the downloader.
+			}()
+
+			// Perform download.
+			chunkIndex, received, err = d.run(chunkIndex, received, f)
+			if err != nil {
+				return errors.New("download iteration did not complete")
```

Comment:
`run` can return severe (i.e. unrecoverable) errors, such as being unable to write to the file handle. It's fine for now; those errors should be very rare.

Instead of a sentinel error, I would prefer to return a `(bool, error)`, where the bool indicates whether another iteration is required. Then the logic below becomes:
```go
if finished {
    break
}
if err != nil {
    // severe error; sleep before trying again
    time.Sleep(time.Second * 90)
}
```